src/utils/deepAssign.js

```diff
@@ -0,0 +1,54 @@
+/*
+Copyright 2019 Adobe. All rights reserved.
+This file is licensed to you under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License. You may obtain a copy
+of the License at http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software distributed under
+the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+OF ANY KIND, either express or implied. See the License for the specific language
+governing permissions and limitations under the License.
+*/
+
+import isObject from "./isObject";
+
+const deepAssignObject = (target, source) => {
+  Object.keys(source).forEach(key => {
+    if (isObject(target[key]) && isObject(source[key])) {
+      deepAssignObject(target[key], source[key]);
+
+      return;
+    }
+
+    target[key] = source[key];
+  });
+};
+
+/**
+ * Recursively copy the values of all enumerable own properties from a source item to a target item if the both items are objects
+ * @private
+ * @param {Object} target - a target object
+ * @param {Object} source - a source object
+ * @example
+ * deepAssign({ a: 'a', b: 'b' }, { b: 'B', c: 'c' });
+ * // { a: 'a', b: 'B', c: 'c' }
+ */
+export default (target, ...sources) => {
+  if (!isObject(target)) {
+    return {};
+  }
+
+  const { length } = sources;
+
+  for (let i = 0; i < length; i += 1) {
+    const source = sources[i];
+
+    if (!isObject(source)) {
+      break;
```

Comment:
We should throw an error here instead of silently failing.  This would happen for `deepAssign({}, "foo")` or `deepAssign({}, undefined)`.

Reply:
Not throwing an error would probably be more native-like, for better or worse. For example:
![image](https://user-images.githubusercontent.com/210820/62646951-b3df1b00-b90c-11e9-95c8-651928f1e136.png)

Reply:
Here's an error though:

![image](https://user-images.githubusercontent.com/210820/62648033-26e99100-b90f-11e9-9d0b-2b5c8de8bffe.png)

Reply:
@jonsnyder and @Aaronius thanks for your suggestions I have added the following changes:
1. If `target` is `null` or `undefined` we will throw a `TypeError`
2. To make sure we are compliant with `Object.assign()` we convert both `target` and `source` to object. Which means that all the weird combinations of different types are supported